builder/amazon/common/ami_config.go

```diff
@@ -54,18 +64,46 @@ func (c *AMIConfig) Prepare(ctx *interpolate.Context) []error {
 				}
 			}
 
+			// Make sure that if we have region_kms_key_ids defined the regions in ami_regions are also in region_kms_key_ids
+			if len(c.AMIRegionKmsKeyIds) > 0 {
+				regionsInKeyMap := make([]string, 0, len(c.AMIRegionKmsKeyIds))
+				for reg := range c.AMIRegionKmsKeyIds {
+					regionsInKeyMap = append(regionsInKeyMap, reg)
+				}
+				if regionsMatch := stringInSlice(region, regionsInKeyMap); !regionsMatch {
+					errs = append(errs, fmt.Errorf("Region %s is in ami_regions but not in region_kms_key_ids", region))
+				}
+			}
+
 			regions = append(regions, region)
 		}
 
 		c.AMIRegions = regions
 	}
+	// Make sure that if we have region_kms_key_ids defined the regions in region_kms_key_ids are also in ami_regions
+	if len(c.AMIRegionKmsKeyIds) > 0 {
+		for kmsKeyRegion := range c.AMIRegionKmsKeyIds {
+			if regionsMatch := stringInSlice(kmsKeyRegion, c.AMIRegions); !regionsMatch {
+				errs = append(errs, fmt.Errorf("Region %s is in region_kms_key_ids but not in ami_regions", kmsKeyRegion))
+			}
+		}
+	}
 
 	if len(c.AMIUsers) > 0 && c.AMIEncryptBootVolume {
 		errs = append(errs, fmt.Errorf("Cannot share AMI with encrypted boot volume"))
 	}
 
-	if len(c.SnapshotUsers) > 0 && len(c.AMIKmsKeyId) == 0 && c.AMIEncryptBootVolume {
-		errs = append(errs, fmt.Errorf("Cannot share snapshot encrypted with default KMS key"))
+	if len(c.SnapshotUsers) > 0 {
+		if len(c.AMIKmsKeyId) == 0 && c.AMIEncryptBootVolume {
+			errs = append(errs, fmt.Errorf("Cannot share snapshot encrypted with default KMS key"))
+		}
+		if len(c.AMIRegionKmsKeyIds) > 0 {
+			for _, kmsKeyRegion := range c.AMIRegionKmsKeyIds {
```

Comment:
the map is region => key id, right? I think naming the value something like kmsKey would be better